Add custom `Qargs` implementation for `Target`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
A custom implementation of `Qargs` was added to `Target` allowing us to directly fix an ownership issue that was currently being solved by `NullableIndexMap`. These additions will ensure the following:
- Seamless conversion from Python `tuple[int]` and Rust `SmallVec<[PhysicalQubit; 2]>`.
- Usage of `&[PhysicalQubit]` as a valid query for the `Target`.

### Details and comments

The structure NullableIndexMap, existed solely to solve an ownership issue that exists with HashMap<K, V> where if K = Option<T> you are unable to query into it by using Option<&T>. To solve these issues we had thought of different workarounds such as improving the `NullableIndexMap` itself (#14195), and removing the `Option` entirely from the mapping in `Target.gate_map` (#14156). Finally, with some suggestions from @jakelishman, we have stumbled onto the following solution. Target will now store a struct called `TargetQargs` in place of `Option<SmallVec<[PhysicalQubit; 2]>>`.

#### `TargetQargs`

This structure represents an enumeration with two different variants:
- `Global` (old case `None`) which is stored when a global operation or a variadic is added.
- `Concrete([PhysicalQubit])` (old case `Some([PhysicalQubit])`) which is stored when a normal operation is added.

It solves the ownership issue on `HashMap` by implementing a `ref` version called `TargetQargsRef` which stores a slice for the `Concrete` variant and implements `Equivalent<TargetQargs>`. This structure also seamlessly converts into `tuple[int]` when it is sent to Python and vice-versa.

### Questions?

Leave a review or a comment in order to discuss! :)

raynelfss committed:
Initial: Add custom `Qargs` implementation for `Target`.

A custom implementation of `Qargs` was added to `Target` allowing us to directly fix an ownership issue that was currently being solved by `NullableIndexMap`. These additions will ensure the following:
- Seamless conversion from Python `tuple[int]` and Rust `SmallVec<[PhysicalQubit; 2]>`.
- Usage of `Option<&[PhysicalQubit]>` as a valid query for the `Target`.

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -14,7 +14,7 @@
 mod errors;
 mod instruction_properties;
-mod nullable_index_map;
+pub mod qargs;
```
From an internal interface perspective, what's the reason for making `qargs` a public module?  Should external packages need to care about this organisation?

(As opposed to publicly exporting `Qargs` from the target root.)

raynelfss replied:
Oh, yes, the original plan was to have `Target` be the one to export this. I will make sure to address this.

raynelfss replied:
done in 544fcb4

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -361,11 +359,11 @@ impl Target {
-        qargs: Option<Qargs>,
+        qargs: TargetQargs,
         properties: Option<InstructionProperties>,
     ) -> PyResult<()> {
```
The logic here looks quite weird - why are `qargs` and `properties` optional?  Certainly it feels _very_ odd that `qargs` should default to `Global`.  I'd expect it to be necessary to specify, but failing that, I'd expect that `None` in Rust space meant "update all".  But really, I don't know why it's optional.

raynelfss replied:
Hmm... You are very much right. This was an outdated assumption of mine. Checking old versions of the code, in no case it was specified that these values would default to `None`. I will work on a fix for this.

raynelfss replied:
Done in 98b7ece

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -406,7 +404,12 @@ impl Target {
+                            .into_pyobject(py)
+                            .unwrap()
+                            .unbind()
+                    })
```
Something's weird about this interface.  I would expect that the iteration item from `qargs_for_operation_name` should already be `TargetQargsRef`, otherwise we're forcing the caller of the function to effectively reconstruct the object.

raynelfss replied:
Done in 544fcb4

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -448,11 +451,11 @@ impl Target {
         py: Python,
-        qargs: Option<Qargs>,
+        qargs: TargetQargs,
     ) -> PyResult<Vec<PyObject>> {
```
Another situation where I'm very confused to why we default to `Global`.

Is there any chance that the defaults were put in because of a PyO3 error saying that "automatic keyword-argument defaults for `Option` are deprecated" or something like that?  If so, the automatic defaults shouldn't have been used in the first place - we should have been using the `signature` field to explicitly say that "these are `Option`, but _don't_ default to `None`".

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -569,7 +572,7 @@ impl Target {
                         // If no qargs operation class is supported
-                        if let Some(_qargs) = &qargs {
+                        if let TargetQargs::Concrete(_qargs) = &qargs {
                             let qarg_set: HashSet<PhysicalQubit> = _qargs.iter().cloned().collect();
```
`_qargs` is used within the body here, so it probably oughtn't to start with an underscore (which denotes "unused").  This is pre-existing, but still.

raynelfss replied:
Fixed in 4abb651e

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -949,14 +926,15 @@ impl Target {
-    /// as pair of `&TargetOperation` and `Option<&Qargs>`.
-    fn _instructions(&self) -> impl Iterator<Item = (&TargetOperation, Option<&Qargs>)> {
+    /// as pair of `&TargetOperation` and `Option<&TargetQargs>`.
+    fn _instructions(&self) -> impl Iterator<Item = (&TargetOperation, TargetQargsRef)> {
```
These comments don't agree with each other.

raynelfss replied:
Fixed in 4abb651e

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -977,8 +955,7 @@ impl Target {
     /// Get the error rate of a given instruction in the target
     pub fn get_error(&self, name: &str, qargs: &[PhysicalQubit]) -> Option<f64> {
```
Why does `get_error` not take `TargetQargsRef`?  Globals can still have errors associated with them.

raynelfss replied:
Fixed in 9602e7c

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -1140,12 +1134,17 @@ impl Target {
         &self,
         operation: &str,
-    ) -> Result<Option<impl Iterator<Item = &Qargs>>, TargetKeyError> {
+    ) -> Result<Option<impl Iterator<Item = &[PhysicalQubit]>>, TargetKeyError> {
```
I would logically expect that the `qargs_for_operation_name` are an iterator over `TargetQargsRef`.  To me, the return type of this function should be `impl Iterator<Item = TargetQargsRef>` - no `Result` (return an empty iterator instead) and no `Option` (`Global` should be part of the iteration).

raynelfss replied:
I think it might be okay to do `&Qargs` which is what I went with in 544fcb4

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -1184,40 +1183,38 @@ impl Target {
         }
-        Some(qargs)
+        Some(qargs.map(|qargs| qargs.as_ref()))
     }
```
This is another place where the outer `None` doesn't make much sense to me.

raynelfss replied:
This I believe referred to mostly to the case in which the only `qarg` available is `None`, and it is how it was done in the past. But I see how this may be a bit confusing since we would ideally want to retuen `Global` alone. Perhaps we need to change the return value to a different `Enum`?.

Either way, this could probably be included in a follow-up PR rather than here.

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+
+use crate::nlayout::PhysicalQubit;
+
+pub type Qargs = SmallVec<[PhysicalQubit; 2]>;
```
I'm not sure a type alias is appropriate here.  When this does leak out to the public, it's necessary to know what the type is, and the alias just masks that, while making it _look_ like a newtype with additional behaviour.

Regardless of whether we keep it or not: from an internal interface perspective, I don't think that this type should get the short name, while the types that users _actually_ have to engage with (`TargetQargs` and `TargetQargsRef`) get long names.  Imo, it'd be more appropriate to flip the names, so users work with `Qargs` and `QargsRef`, and the verbose internal storage type is confined to `Target` internals.

raynelfss replied:
Fixed in 8d3e09d

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+    #[default]
+    Global,
+    Concrete(Qargs),
+}
```
What's the reason for defining a `Default`?   That seems like highly surprising behaviour to me, especially given how poorly `Global` is handled by our codebase overall.

I'm also not clear what semantics `PartialOrd` and `Ord` are supposed to have.  I can _just about_ see how `PartialOrd` would be defined within `Concrete`, but I don't know _why_ you should ever want that, and I don't think that `Global` has any clear well-defined ordering with respect to `Concrete`.

raynelfss replied:
Fixed in 9602e7c

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+    pub fn as_ref(&self) -> TargetQargsRef<'_> {
+        match self {
+            TargetQargs::Global => TargetQargsRef::Global,
+            TargetQargs::Concrete(small_vec) => TargetQargsRef::Concrete(small_vec),
```
Minor point of style: it's usually not a great idea to name a variable (`small_vec`) based on its type.  For one, it's simply redundant, because any competent IDE can easily _tell_ you the type.  For two, it's liable to become out-of-date or incorrect, and it means that we're not naming the variable with what it represents.  Even here, it's already incorrect; `small_vec` names a reference pattern that actually matches a type `&[PhysicalQubit]`.  The bound variable on the RHS is not a `SmallVec`.

raynelfss replied:
Oh yeah, here vs code generated the names when usong the `match` autocomplete from `rustfmt`.

raynelfss replied:
Fixed in 9602e7c

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+            Self::Concrete(small_vec) => Ok(PyTuple::new(py, small_vec)?.into_any()),
+        }
+    }
+}
```
A large reason to swap from `Option<[PhysicalQubit]>` to `enum Qargs { Global | Concrete([PhysicalQubit]) }` is because the semantics of `Option` (value, or absence of value) are not the same.  These methods are mapping back to the incorrect semantics.  While it's true that there's a mathematical bijection between the two types, that doesn't mean that the semantics are the same.  `Global` is not the "absence" of `Qargs`, it's a proper value of `Qargs` in itself.

`Option::iter` exists because `Option<T>` semantically represents either 0 or 1 values of type `T`.  `Qargs` always represents _exactly_ 1 value of `Qargs`.  `unwrap` similarly asserts that the value is present, but `Global` _is_ a value.  `unwrap_or_default` is particularly problematic, because not only does it have the bad semantics of `unwrap`, but additionally it now _modifies_ the (already questionable) default of `Qargs` to be the 0-length vector instead of `Global`, so now there's a clash.

I would remove all of these highlighted methods.

raynelfss replied:
I agree with removing all of the `impl` methods highlighted, but why should we remove the `IntoPyObject` implementation here? Since this is an internal structure meant mainly for usage in Rust, I don't think it is a good idea to remove the conversion into the `Python` types that represent the old values. We have not exposed this as a pyo3 struct yet, nor did I believe we were going to.

raynelfss replied:
Fixed partially in 9602e7c

jakelishman replied:
Oh the python impl being highlighted is a mistake - I must have dragged too far.

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+    #[default]
+    Global,
+    Concrete(&'a [PhysicalQubit]),
+}
```
Similar comment here about `Default`, `PartialOrd` and `Ord`.

raynelfss replied:
Fixed in 9602e7c

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+            Self::Global => &[],
+            Self::Concrete(qargs) => qargs,
+        }
+    }
```
Same comment about these `Option` methods.

raynelfss replied:
Fixed in 9602e7c

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+                TargetQargs::Concrete(physical_qubits.to_owned().into())
+            }
+        }
+    }
```
This isn't perfect (should be the trait impl), but I see the problem with `Borrow`.  I suspect this means that there's a better way to do the modelling of `Qargs`, but I'm not immediately clear what it is.

raynelfss replied:
Exploring...

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+            None => Self::Global,
+        }
+    }
+}
```
This `impl` is convenient for coming from Python, though I'm not _super_ wild about it in Rust.  It's probably fine to keep, but we should probably only use it for Python.

raynelfss replied:
Removed in 9602e7c

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+    fn from(value: &'a [PhysicalQubit]) -> Self {
+        Self::Concrete(value)
+    }
+}
```
We should define our interfaces in `Target` so that users only ever really construct `&[PhysicalQubit]`, and `Target` uses generics to turn it into a `TargetQargsRef`.

raynelfss replied:
I like the idea of using generics here. The only situation in which it may not work as intended is whenever we use `Index<T>` for the gate_map. As it will expect a `TargetQarg` or an `Equivalent`.

raynelfss replied:
Added in 544fcb4

jakelishman commented on crates/accelerate/src/target_transpiler/qargs.rs:
```diff
@@ -0,0 +1,278 @@
+            }
+        }
+    }
+}
```
Same comment about iterators.

raynelfss replied:
Removed in 9602e7c

jakelishman commented on crates/accelerate/src/unitary_synthesis.rs:
```diff
@@ -114,7 +116,7 @@ fn get_euler_basis_set(basis_list: IndexSet<&str>) -> EulerBasisSet {
 fn get_target_basis_set(target: &Target, qubit: PhysicalQubit) -> EulerBasisSet {
     let mut target_basis_set: EulerBasisSet = EulerBasisSet::new();
-    let target_basis_list = target.operation_names_for_qargs(Some(&smallvec![qubit]));
+    let target_basis_list = target.operation_names_for_qargs(TargetQargsRef::Concrete(&[qubit]));
```
When I'm talking about the generics in the interfaces, what I mean is that it should be possible to specify _either `TargetQargsRef::Concrete(&[qubit])` here, or simply `&[qubit]`, and both should work.

raynelfss replied:
Added in 9602e7c

jakelishman commented on crates/accelerate/src/unitary_synthesis.rs:
```diff
@@ -827,11 +829,9 @@ fn preferred_direction(
-                                    )) {
+                                    Ok(_) => match target[decomposer.packed_op.name()]
+                                        .get::<TargetQargsRef>(&Some(q_tuple.as_slice()).into())
+                                    {
```
Something looks very funky with the typing here.

raynelfss replied:
Indeed... Fixed in 9602e7c

jakelishman commented on crates/accelerate/src/vf2_layout.rs:
```diff
@@ -53,15 +53,17 @@ create_exception!(qiskit, MultiQEncountered, pyo3::exceptions::PyException);
             }
         }
+        // All qargs accessed by this function are guaranteed to be concrete.
+        let qargs = qargs.unwrap();
```
When you're using a strong type system like Rust, the aim is to encode impossible states directly into the type system, so you don't need thing like the "unwrap" or the comments "guaranteed to be concrete" wherever possible.

In this case, the way you do that is by making sure that the iterator isn't "an iterator over `TargetQargsRef` except I pinky swear that none of them will be globals", it's to make it "an iterator over `&[PhysicalQubit]` items" (or use `if let` without the loop body to do the same thing).  So you do things like
```rust
for qargs in target.qargs().flatten() {
    let TargetQargsRef::Concrete(qargs) = qargs else {
        continue;
    };
}
```
then you've encoded the guarantee into the type system, so you never have to _assume_ it, it's statically proven.  (Although in my universe, the `.flatten()` wouldn't be necessary either.)

(Same comment elsewhere in this file.)

raynelfss replied:
Fixed in 9602e7c

jakelishman reviewed (commented):
Thanks for looking at this Ray.  I think changing the `Option<Vec<PhysicalQubit>>` (or whatever) to the explicit enum certainly gives us a way to have a lot cleaner modelling in the interface.  I think this PR's gone some of the way, but there's still a few places in it where the old system is persisting, or places where it's still a bit sharp to use.

Some recurring comments I saw:
- In a lot of places, we use have Python-space functions default `qargs` to `Global`.  In at least some of these cases, the original Python-space `Target` didn't have this defaulting, and certainly defaulting to `Global` to me seems to generally be unexpected behaviour.
- In all places that we can control (i.e. all `Target` methods), the user probably shouldn't ever be required to give a `TargetQargsRef`.  It would be nicer if such methods instead took `T: Into<TargetQargsRef>`, then we can have a short-hand that allows looking up with a slice.  For example, something like
  ```rust
  impl<'a> From<&'a [PhysicalQubit]> for TargetQargsRef<'a> {
    fn from(val: &'a [PhysicalQubit]) -> Self {
      Self::Concrete(val)
    }
  }

  struct Target {}
  impl Target {
    pub fn get_error<'a, T>(&'a self, qargs: T) -> Option<f64>
    where
      T: Into<TargetQargsRef<'a>>
    {}
  }
  ```
  (I didn't compile that, but I generally just mean the principle.)

raynelfss committed:
Refactor: `TargetQargs` is now `Qargs`.
- `TargetQargsRef` is now `QargsRef`.

raynelfss committed:
Fix: Use generics for querying into the `Target`.
 Use a `pub use` to export `Qargs`, and `QargsRef` directly from the `target_transpiler` module.
 - Add additional conversion to `&Qargs` to `PyTuple`.
 - Add `From<[T; N]>` for `Qargs` and `QargsRef`.
 - Enable use of generics to query into the `Qargs` present in `Target`. These are based on `Into<QargRef>` and include `&Qarg`, `&[PhysicalQubit]` and others.

raynelfss committed:
Fix: Make `Qargs` and `QargsRef` more lean
- Remove unnecessary methods from `Qargs` as pointed out by @jakelishman.
- Remove macro derives `Default`, `Ord`, and `PartialOrd` as they had incorrect semantic implications.
- Added more generic `From` and `FromIterator` implementations to make usage more versatile.

raynelfss commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -547,19 +550,19 @@ impl Target {
     #[pyo3(
         name = "instruction_supported",
-        signature = (operation_name=None, qargs=None, operation_class=None, parameters=None)
+        signature = (operation_name=None, qargs=TargetQargs::Global, operation_class=None, parameters=None)
```
Do you think thiw syntax is valid here, it feels strange that qargs is global by default here? The python counterpart did have `None` as a default for everything, but for this we might prefer to use `Option<Qargs>` but I'm not quite sure.

raynelfss committed:
Fix: `update_instruction_properties` signature.

raynelfss committed:
Fix: update signatures for `for_qargs` methods.

raynelfss committed:
Refactor: Remove wrongful use of underscore prefixed names.
- Fix docstrings documenting `Qargs` as `TargetQargs`.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-qargs

jakelishman commented on crates/accelerate/src/basis/basis_translator/mod.rs:
```diff
@@ -133,37 +135,45 @@ fn run(
-                .cloned()
-                .collect();
+            // `extract_basis_target` should never add a 'Global' qarg into this mapping.
+            unreachable!("Attempted expanding the target with a 'Global' qarg.")
```
A couple of things here:
1. when we've got
   ```rust
   if let patt = scrutinee {
       // ... body ...
   } else {
       unreachable!("...");
   }
   ```
   (or any diverging code, like a `break`, `continue`, `return` or any panic), we can reduce a level of nesting and localise the error handling by doing
   ```rust
   let patt = scrutinee else {
       unreachable!("...");
   };
   // ... body ...
   ```
   which is usually a fair amount more readable, especially when we have to do it a couple of times
2. if `extract_basis_target` _should_ never add a `Global` qarg into the mapping, then can we make the mapping `IndexMap<Vec<Qubit>, T>` instead of `IndexMap<Qargs, T>` so it's enforced in the type system?

raynelfss replied:
Addressed in 4006e725879580da3bff7a7c15035216ed7f9bc6

jakelishman commented on crates/accelerate/src/gate_direction.rs:
```diff
@@ -225,7 +225,8 @@ fn py_fix_direction_target(
         }
-        target.instruction_supported(inst.op.name(), Some(&qargs))
+        // This will come back to bite me :)
+        target.instruction_supported(inst.op.name(), &qargs)
```
why so?

raynelfss replied:
Forgot to remove that comment, but in case you want to know if it did, **it absolutely did**.

raynelfss replied:
Removed in b010407f2b40d518a210e4c5bd273908b637a3cf

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -23,15 +23,16 @@ use ahash::RandomState;
     IntoPyObjectExt,
 };
 
+pub use qargs::{Qargs, QargsRef};
```
very nitpicky: can we move the `pub use` stuff somewhere that's not hiding in the middle of the private `use` statements?

raynelfss replied:
Moved in 4006e725879580da3bff7a7c15035216ed7f9bc6

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -398,17 +395,9 @@ impl Target {
-            })),
+            Ok(option_set) => Ok(option_set.map(|qargs| qargs.collect())),
             Err(e) => Err(PyKeyError::new_err(e.message)),
         }
```
The documented return type of this function and the actual return type don't seem to match for two reasons: the comment says it should be a set, but it's actually a list, and the actual return value can sometimes be `None`.  I haven't checked when the discrepancy was introduced, so I don't know if the trouble is in the comment or the function.

raynelfss replied:
Actually, I don't think this function is being used currently, as the Python version differs from this. I will change this docstring, but the correct usage will be added in #14173.

raynelfss replied:
Addressed partially in 4006e725879580da3bff7a7c15035216ed7f9bc6

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -549,19 +534,19 @@ impl Target {
     ///     bool: Returns ``True`` if the instruction is supported and ``False`` if it isn't.
     #[pyo3(
         name = "instruction_supported",
-        signature = (operation_name=None, qargs=None, operation_class=None, parameters=None)
+        signature = (operation_name=None, qargs=Qargs::Global, operation_class=None, parameters=None)
     )]
     pub fn py_instruction_supported(
```
The documentation actually says that if `qargs` isn't provided, then the function should return true if it's supported on _any_ qargs, not necessary global.  We might have a testing gap if this default is passing.

We can potentially fix it in a follow-up, if there _is_ a problem, but I think a true fix might be slightly involved and require some assistance from Python space, because we can't use `None` as the sentinel for "not passed".

jakelishman commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -1177,33 +1160,35 @@ impl Target {
-            if let Some(_qargs) = qargs {
-                let qarg_set: HashSet<&PhysicalQubit> = _qargs.iter().collect();
+            if let QargsRef::Concrete(qargs_as_vec) = qargs {
+                let qarg_set: HashSet<&PhysicalQubit> = qargs_as_vec.iter().collect();
```
This could likely benefit from a switch to `let QargsRef::Concrete(qargs) = qargs else { return None; };`

raynelfss replied:
Addressed in 4006e725879580da3bff7a7c15035216ed7f9bc6

jakelishman reviewed (commented):
Thanks for the changes Ray - this really is looking a lot neater now.

raynelfss committed:
Fix: Address review comments
- `extract_target_basis` uses `SmallVec` instead of `Qargs`.
- Use`let Some(val) = conscript else { // }` pattern.
- Fix docstring for `py_qargs_for_operation_name`.
- Move `pub use` of qargs to top of document.

raynelfss committed:
Docs: Remove stray comment

jakelishman reviewed (approved):
Thanks for the changes!
